Csv export 333

wgilling opened the pull request:
This feels pretty big and the likelihood that it needs more work on some fields is pretty high in my estimation, but it's time to get the merge process started with a review.

First note is that all multi-field delimiters have been changed from the "||" to " || " so that items with missing parts such as no subtitle on a complex title paragraph may be imported without confusion. The migrate_plus.migration.csv.yml has this delimiter coded into those fields, so I only mention this because the previous migrations from legacy CSV had the delimiters with no spaces around them.

The view `rest_csv_export` provides three different displays -- one for single item, collection, and search. These all have endpoints that make it fairly easy to manually download any given CSV export. For example, and item that is at `items/12345` has the route `items/12345/csv` and a collection that has `collections/123` has the csv route at `collections/123/csv` and any search pattern should be supported by adding the /csv to the end such as `search/csv?search_api_fulltext=data`. There should be a link in the Admin Toolbox to Download CSV from the items and collections page and another link on search pages near the "Export MODS XML".

The code uses Migrate Source UI in order to allow for a web interface for importing any CSV that was exported and modified. This is at http://localhost:8000/migrate_source_ui and is "Import/Update nodes from CSV (supports csv)".

To test, it seems that you know what to do... but it can be a fairly long process to set up alternate field vocabulary for testing purposes. When exporting, I found that it saves a lot of time to use a collection that has less than 5 items... not that the export takes a long time to run, but it is much easier to inspect and edit a smaller CSV file.

The import csv configurations will need to be imported. These are at `web/modules/custom/asu_migrate/config/install/migrate_plus.migration.csv.yml`.

For a command-line import test, the CSV file that is stored at web/modules/custom/asu_migrate/data/csv.csv can be overwritten in order to just use the `drush migrate:import csv --update` command.

Any questions, please ask.

wgilling committed:
added both views_data_export and entity_export_csv for evaluation, made a test view to export content as CSV for #333

wgilling committed:
wrote a TaxonomyWithURI field formatter for the view output and configured the test view to use this for several fields

wgilling committed:
renamed the csv export view

wgilling committed:
moved formatter plugin into asu_migrate module

wgilling committed:
removed entity_export_csv

wgilling committed:
getting errors

wgilling committed:
replaced noted-content view with a solr view so a collection export of all csv can take advantage of the Ancestors field for #333

wgilling committed:
added typed relators formatter with selection of agent type to output Person Contributor, Corporate Contributor, Event Contributor

wgilling committed:
added complex title for CSV formatter and reworked subjects fields on view

wgilling committed:
added Table of Contents, Genre, Language, Extent, Preferred Citation, Copyright Statement, Reuse Permissions, and Handle for #333

wgilling committed:
added ID, Legacy System ID, System Creator, Peer-reviewed, Open Access, Series and adjusted view to export for single items as well as collections for #333

wgilling committed:
added Date Created, Copyright Date, Date Digitized, Edition, Issuance, Frequency, Related Item, and Related Item Type for #333

wgilling committed:
added Parent Type formatter to handle Parent collection vs. complex object parent for #333

wgilling committed:
added Visibility formatter and fields: Additional Memberships, Place of Publication Text, Place of Publication Code, Coordinates, Note - a work in progress for #333

wgilling committed:
added formatter for complex note paragraphs, adjusted view accordingly for #333

wgilling committed:
cleaned up Complex Title CSV formatter and added Typed identifier for CSV formatter for #333

wgilling committed:
added Cataloging Standards, Description Source, Level of Coding, Derivative File Permissions, Original File Permissions, and Embargo Release Date for #333

wgilling committed:
adjusted the Parent Type CSV formatter to only return Parent Collection for items that are not complex object children, adjusted search_api processor for institution_from_collection to recurse so that complex object children could have the right value for #333

wgilling committed:
round-trip CSV import with some issues to work out... some delimiter parsing not working correctly in some cases for #333

wgilling committed:
some development for fields that are exported but not coded into the importer yet for #333

wgilling committed:
export created date in Y-m-d G:i:s.uP format because it did not like the default short date format, view now supports CSV exports from search page for #333

wgilling committed:
adjusted the ParagraphGenerate class to delimit on a given character, bugfix to the output to match the legacy delimiting using : char for #333

wgilling committed:
added link to admin toolbox and on search form to download CSV, added placeholder formatter for importing Geolocation field values, importing Level of Coding, Statment of Responsibility, Complex object child, Issuance, Frequency, Edition, Open access, Peer-reviewed, Title Subject, Temporal Subject, possibly more for #333

wgilling committed:
fixed str_to_geolocation processor for #333

wgilling committed:
trying to prevent encoding of the title value in export for #333

wgilling committed:
made each csv export display have a corresponding download filename item_{date}.csv, collection_{date}.csv, and search_{date}.csv... node id is not available as a token here. for #333

wgilling committed:
adjust the NameURILookup plugin to limit results on the provided default_vocabulary (optional) and minor adjustment to the import CSV script for #333

wgilling committed:
add import script for Place of Pub Code, Place of Pub Text, Cataloging Standards, Deriv file perms, Orig file perms, Description source, and Embargo release date for #333

wgilling committed:
updated the import script destination | overwrite_properties to include all that are imported for #333

wgilling committed:
migrate_devel and migrate_source_ui modules enabled for #333

elizoller commented on web/modules/custom/asu_migrate/src/Plugin/Field/FieldFormatter/VisibilityCSVFormatter.php:
```diff
@@ -0,0 +1,41 @@
+    //  - Public: published
+    foreach ($items as $delta => $item) {
+      $item_entity_id = $item->value;
+      $item_entity = @\Drupal::entityTypeManager()->getStorage('node')->load($item_entity_id);
```
dependency injection?

elizoller commented on web/modules/custom/asu_admin_toolbox/src/Plugin/Block/AdminToolboxBlock.php:
```diff
@@ -141,7 +142,11 @@ public function build() {
     $is_complex_object = FALSE;
-    if ($node->bundle() == 'asu_repository_item') {
+    $is_asu_repository_item = ($node->bundle() == 'asu_repository_item');
+    $user = User::load(\Drupal::currentUser()->id());
```
dependency injection?

elizoller commented on web/modules/custom/asu_migrate/config/install/migrate_plus.migration.c270_node.yml:
```diff
@@ -0,0 +1,522 @@
+# Uninstall this config when the feature is uninstalled
```
can we take this file out of this PR?

Commit:
attempting dependency injection for #333

wgilling committed:
fixed dependency injection after testing for #333

wgilling commented:
Addressed the dependency injection issues and removed the c270 file. There were four process plugins that actually needed entityTypeManager to use dependency injection. There was also one bit in the AdminToolboxBlock code that needed dependency injection to replace the User:: call.

Please let me know if there are any issues when reviewing.

wgilling commented:
Re-ready to re-review.

elizoller committed:
merge develop

elizoller committed:
adding back in removed modules and adding some NULL entity checks

elizoller committed:
removing duplicate line which didn't account for null values

elizoller commented on web/modules/custom/asu_migrate/config/install/migrate_plus.migration.csv.yml:
```diff
@@ -0,0 +1,513 @@
+    source: Extent
+    method: process
+
+# wgg - Date Created should not be the EDTF field.
```
why is this commented out?

elizoller commented on web/modules/custom/asu_migrate/src/Plugin/Field/FieldFormatter/VisibilityCSVFormatter.php:
```diff
@@ -0,0 +1,80 @@
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $elements = [];
+    $asu_utils = \Drupal::service('asu_utils');
```
:D dependency injection?

elizoller reviewed (changes requested):
I tested the export on several levels (search, collection, item) and after the minor tweaks i pushed up, it seems ok for the collections i have in my local. 

i haven't tested the import back in feature, but did see one thing that'll need addressing which i flagged (the missing date created field)

elizoller committed:
styling audio/video controls

elizoller committed:
remove the poster for audio

elizoller committed:
more tweaks with audio/video display

elizoller committed:
any luck on the title csv formatteR?

wgilling committed:
restored Created date for CSV import handling, also reworked to use dependency injection for an asu_utils method call

wgilling commented:
The latest commit was tested to export and import Created Date and correctly update the node value for that EDTF field.

elizoller committed:
merge develop

elizoller committed:
Merge branch 'csv_export_333' of https://github.com/asulibraries/islandora-repo into csv_export_333
